site/content/entry/improvements-to-hardware-renderer.md

```diff
@@ -0,0 +1,200 @@
+On the PICA200, `infinity * 0 = 0` but in OpenGL `infinity * 0 = NaN` and this can't be configured.
+In the generated GLSL shaders, phantom emulates this behavior by making a function call instead of a simple multiplication.
+
+```glsl
+vec4 sanitize_mul(vec4 lhs, vec4 rhs) {
+    vec4 product = lhs * rhs;
+    return mix(product, mix(mix(vec4(0.0), product, isnan(rhs)), product, isnan(lhs)), isnan(product));
+}
+```
+
+Alas, it's a performance penalty to use a function everywhere instead of regular multiplication.
+On weaker GPUs, we noticed the penalty is so severe, we actually made this configurable.
+The whole point of a hardware renderer is to be fast, so eating a penalty when only a small handful of games need this level of accuracy would be regrettable.
+You can turn off this feature in the settings by deselecting "Accurate Shader" and get a noticeable performance boost, but be aware that a few games will break in strange ways!
+
+### Finding Bugs and Working Overtime
+
+We were very excited to launch this feature when phantom declared that it was ready; results from user testing was entirely positive, and the performance improvements were unbelievable, but one thing stood in the way.
+No one had yet tested to see if it worked on AMD GPUs.
+We called for our good friend [JMC47](https://dolphin-emu.org/blog/authors/JMC47/) to break out the AMD card he uses for testing Dolphin, and Citra crashed the driver! Oh no!
+
+From JMC47's time in Dolphin, he's made a few friends here and there, and he found someone willing to investigate.
+After a few grueling weeks, this developer was able to narrow down what the problem is, and luckily it's not a bug in the AMD drivers.
+It turns out that it's a bug in the GL specification, or more precisely, the exact issue is ambiguous wording.
+[glDrawRangeElementsBaseVertex](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDrawRangeElementsBaseVertex.xhtml) states that the indices should be a pointer, but doesn't say whether the pointer should be to CPU memory or GPU memory.
+Citra passed a pointer to CPU memory without a second thought, as both Nvidia and Intel drivers seemed fine with it, but AMD drivers are strict.
+As a workaround, phantom added support for streaming storage buffers, which allows Citra to work with the data on the CPU and flush it to the GPU when it's time to draw.
+
+It's a challenge to support all of the many GPUs out there, and we've put in so much work to ensure that this new feature will run on as many hardware configurations as possible.
+But it's very likely that there will be some GPUs that do not fully support the new hardware renderer, and so we added another option in the Configuration to allow users to turn this feature off entirely.
+Simply changing the "Hardware Renderer" from "GPU Shaders" to "CPU Shaders" will revert to using the same old CPU JIT shaders that Citra was using before.
```

Comment:
Worth stating what the new default behavior will be, IMO
Also, worth adding a screenshot of the config as well with the setting highlighted, IMO